File: src/lib/fx/emitter/cpp/emitter.cpp
#pragma once 

#include <stdio.h>

#include <cassert>
#include <vector>
#include <iostream>
#include <algorithm>
#include <sstream>
#include <memory>
#include <utility>

#include "emitter.h" 

namespace IFX
{

namespace FxTranslator
{
    // attention! must be synced with FxTranslator.ts (!)
    static std::string UAV_PARTICLES = "uavParticles";
    static std::string UAV_STATES = "uavStates";
    static std::string UAV_DEAD_INDICES = "uavDeadIndices";
    static std::string UAV_CREATION_REQUESTS = "uavCreationRequests";
    static std::string UAV_PRERENDERED = "uavPrerendered";
    static std::string UAV_SERIALS = "uavSerials";
    static std::string UAV_SPAWN_DISPATCH_ARGUMENTS = "uavSpawnDispatchArguments";
}

std::string UavSerials(uint32_t i)
{
    std::ostringstream s;
    s << FxTranslator::UAV_SERIALS << i;
    return s.str();
}

std::string UavPrerendered(uint32_t i)
{
    std::ostringstream s;
    s << FxTranslator::UAV_PRERENDERED << i;
    return s.str();
}


std::string UavPrerenderedSorted(uint32_t i)
{
    std::ostringstream s;
    s << FxTranslator::UAV_PRERENDERED << i << "Sorted";
    return s.str();
}


VM::BUNDLE_UAV CreateUAVEx(const Fx::UAVBundleT &bundle, int capacity)
{
    return VM::BUNDLE::CreateUAV(bundle.name, bundle.stride, capacity, bundle.slot);
}


std::vector<VM::BUNDLE_UAV> CreateUAVsEx(const std::vector<std::unique_ptr<Fx::UAVBundleT>> &bundles, int capacity, std::vector<VM::BUNDLE_UAV> &sharedUAVs)
{
    std::vector<VM::BUNDLE_UAV> uavs;
    for (auto &uavBundle : bundles)
    {
        auto sharedUAV = find_if(begin(sharedUAVs), end(sharedUAVs),
                                 [&uavBundle](const VM::BUNDLE_UAV &el)
                                 { return el.name == uavBundle->name; });
        if (sharedUAV != end(sharedUAVs))
        {
            uavs.push_back(*sharedUAV);
            continue;
        }
        VM::BUNDLE_UAV uav = CreateUAVEx(*uavBundle, capacity);
        uavs.push_back(uav);
        sharedUAVs.push_back(uav);
    }
    return uavs;
}


BYTECODE_BUNDLE SetupFxRoutineBytecodeBundle(
    std::string debugName,
    const Fx::RoutineBytecodeBundleT &routineBundle,
    int capacity,
    std::vector<VM::BUNDLE_UAV> &sharedUAVs)
{
    auto uavs = CreateUAVsEx(routineBundle.resources->uavs, capacity, sharedUAVs);
    
    VM::BUNDLE vmBundle(debugName, VM::memory_view::FromVector(routineBundle.code));
    VM::BUNDLE_NUMTHREADS numthreads{
        routineBundle.numthreads[0],
        routineBundle.numthreads[1],
        routineBundle.numthreads[2]
    };

    for (auto &uav : uavs)
    {
        vmBundle.SetInput(uav.index, uav.buffer);
    }

    BYTECODE_BUNDLE bcBundle;
    bcBundle.uavs = std::move(uavs);
    bcBundle.vmBundle = std::move(vmBundle); // << important to copy because of circular dependencies of constants
    bcBundle.numthreads = numthreads;
    return bcBundle;
}


EMITTER_PASS::EMITTER_PASS (
    const EMITTER* pParent, 
    uint32_t id, 
    const EMITTER_DESC& desc, 
    const BYTECODE_BUNDLE& bundle
)
    : m_parent(pParent)
    , m_id(id)
    , m_desc(desc)
    , m_prerenderBundle(bundle)
{

}


VM::BUNDLE_UAV* EMITTER_PASS::UavNonSorted() { return const_cast<VM::BUNDLE_UAV*>(Parent().Uav(UavPrerendered(m_id))); }
VM::BUNDLE_UAV* EMITTER_PASS::UavSorted() { return const_cast<VM::BUNDLE_UAV*>(Parent().Uav(UavPrerenderedSorted(m_id))); }

const VM::BUNDLE_UAV* EMITTER_PASS::UavNonSorted() const { return Parent().Uav(UavPrerendered(m_id)); }
const VM::BUNDLE_UAV* EMITTER_PASS::UavSorted() const { return Parent().Uav(UavPrerenderedSorted(m_id)); }

const VM::BUNDLE_UAV* EMITTER_PASS::UavSerials() const { return Parent().Uav(IFX::UavSerials(m_id)); }
const VM::BUNDLE_UAV* EMITTER_PASS::UavStates() const { return Parent().Uav(FxTranslator::UAV_STATES); }

const EMITTER& EMITTER_PASS::Parent() const
{
    return *m_parent;
}


uint32_t EMITTER_PASS::GetNumRenderedParticles() const
{ 
    return Parent().GetNumParticles() * m_desc.instanceCount; 
}


VM::memory_view EMITTER_PASS::GetData() const
{
    return UavSorted()->data;
}


void EMITTER_PASS::Serialize()
{
    // NOTE: yes, I understand this is a crappy and stupid brute force sorting,
    //       I hate javascript for that :/

    uint32_t nStride = m_desc.stride * m_desc.instanceCount; // stride in floats

    assert(UavSorted()->data.size == nStride * Parent().GetCapacity());

    auto* src = UavNonSorted()->data.As<float_t>();
    auto* dst = UavSorted()->data.As<float_t>();

    auto* states = UavStates()->data.As<int32_t>();
    auto* serials = UavSerials()->data.As<int32_t>();

    std::vector<std::pair<uint32_t, int32_t>> indicies;

     for (uint32_t iPart = 0; iPart < Parent().GetCapacity(); ++iPart) 
     {
        if (states[iPart]) 
        {
            indicies.push_back({ iPart, GetDesc().sorting ? serials[iPart * GetDesc().instanceCount] : 0 });
        }
    }
 
    if (GetDesc().sorting)
    {
        std::sort(begin(indicies), end(indicies), 
            [](const std::pair<uint32_t, uint32_t>& a, const std::pair<uint32_t, uint32_t>& b) {
                return a.second > b.second; 
            });
    }

    for (uint32_t i = 0; i < indicies.size(); ++i) {
        uint32_t iFrom = indicies[i].first * nStride;
        uint32_t iTo = i * nStride;
        memcpy(dst + iTo, src + iFrom, nStride * sizeof(float_t));
    }
}



void EMITTER_PASS::Prerender(const UNIFORMS& uniforms) 
{
    auto& bundle = m_prerenderBundle;
    assert(Parent().GetCapacity() % bundle.numthreads.x == 0);

    auto uav = find_if(begin(bundle.uavs), end(bundle.uavs), 
        [&](const VM::BUNDLE_UAV& uav) { return uav.name == UavPrerendered(m_id); });
    assert(uav != end(bundle.uavs));
    uav->OverwriteCounter(0);
    bundle.SetConstants(uniforms);
    bundle.Run(Parent().GetCapacity() / bundle.numthreads.x);
}


// struct DefaultShaderInput {
//     vec3 pos;
//     vec4 color;
//     float size;
// };

void EMITTER_PASS::Dump() const 
{
    auto npart = GetNumRenderedParticles(); 
    std::cout << "particles rendered total: " << npart << std::endl;

    for (int iPart = 0; iPart < npart; ++ iPart)
    {
        // auto& part = UavNonSorted()->data.As<DefaultShaderInput>()[iPart];
        // std::cout << "part(" << iPart 
        // << ") = { size: " << part.size 
        // << ", pos: " << part.pos.x << ", " << part.pos.y << ", " << part.pos.z 
        // << " } " << std::endl;
    }
}


EMITTER::EMITTER(void* buf)
{
    Fx::BundleT fx;
    const Fx::Bundle *pBundle = Fx::GetBundle(buf);
    pBundle->UnPackTo(&fx);
 
    std::cout << "==========================================" << std::endl;
    std::cout << "   bundle name: " << fx.name << std::endl;
    std::cout << "bundle version: " << fx.signature->version << std::endl;
    std::cout << "==========================================" << std::endl;

    auto [name, signature, meta, content] = fx;
    auto [capacity, simulationRoutines, renderPasses, particle] = *content.AsPartBundle();

    m_name = name;
    m_capacity = capacity;
    m_particle = Fx::TypeLayoutT(*particle); // create new copy of type layout
    
    auto& sharedUAVs = m_sharedUAVs;
    m_resetBundle = SetupFxRoutineBytecodeBundle(
        name + "/reset",
        *simulationRoutines[Fx::EPartSimRoutines_k_Reset].AsRoutineBytecodeBundle(),
        capacity,
        sharedUAVs);
    m_initBundle = SetupFxRoutineBytecodeBundle(
        name + "/init",
        *simulationRoutines[Fx::EPartSimRoutines_k_Init].AsRoutineBytecodeBundle(),
        capacity,
        sharedUAVs);
    m_updateBundle = SetupFxRoutineBytecodeBundle(
        name + "/update",
        *simulationRoutines[Fx::EPartSimRoutines_k_Update].AsRoutineBytecodeBundle(),
        capacity,
        sharedUAVs);
    m_spawnBundle = SetupFxRoutineBytecodeBundle(
        name + "/spawn",
        *simulationRoutines[Fx::EPartSimRoutines_k_Spawn].AsRoutineBytecodeBundle(),
        4,
        sharedUAVs);

    auto& passes = m_passes;
    for (int i = 0; i < renderPasses.size(); ++i)
    {
        auto &pass = renderPasses[i];
        auto [routines, geometry, sorting, instanceCount, stride, instance] = *pass;

        const Fx::RoutineBytecodeBundleT &prerender = *routines[Fx::EPartRenderRoutines_k_Prerender].AsRoutineBytecodeBundle();
        BYTECODE_BUNDLE bundle = SetupFxRoutineBytecodeBundle(
            name + "/prerender",
            prerender,
            capacity * instanceCount,
            sharedUAVs);

        auto uav = find_if(begin(bundle.uavs), end(bundle.uavs), [&](const VM::BUNDLE_UAV &uav)
                            { return uav.name == UavPrerendered(i); });

        std::string vertexShader = routines[Fx::EPartRenderRoutines_k_Vertex].AsRoutineGLSLBundle()->code;
        std::string pixelShader = routines[Fx::EPartRenderRoutines_k_Pixel].AsRoutineGLSLBundle()->code;

        // note: only GLSL routines are supported!
        std::vector<std::unique_ptr<Fx::GLSLAttributeT>> &attrs = routines[Fx::EPartRenderRoutines_k_Vertex].AsRoutineGLSLBundle()->attributes;

        auto &prerenderUAVs = prerender.resources->uavs;
        auto uavPrerendReflect = std::find_if(std::begin(prerenderUAVs), std::end(prerenderUAVs), [&](const std::unique_ptr<Fx::UAVBundleT> &uav)
                                            { return uav->name == UavPrerendered(i); });

        {
            std::vector<std::unique_ptr<Fx::UAVBundleT>> bundles;
            bundles.push_back(std::make_unique<Fx::UAVBundleT>(**uavPrerendReflect));
            bundles.back()->name = UavPrerenderedSorted(i);

            CreateUAVsEx(bundles, capacity * instanceCount, sharedUAVs)[0];
        }

        {
            EMITTER_DESC desc;
            desc.geometry = geometry;
            desc.sorting = sorting;
            desc.stride = stride;
            desc.instanceCount = instanceCount;
            desc.renderInstance = Fx::TypeLayoutT(*instance);

            for (auto& attr : attrs)
            {
                desc.instanceLayout.push_back({ attr->size, attr->offset, attr->name });
            }
            
            desc.vertexShader = vertexShader;
            desc.pixelShader = pixelShader;

            passes.emplace_back(this, i, desc, bundle);
        }
    }
}


EMITTER::~EMITTER()
{
    for (auto& uav : m_sharedUAVs)
    {
        VM::BUNDLE::DestroyUAV(uav);
        // std::cout << "UAV '" << uav.name << " has been destroyed." << std::endl;
    }
    std::cout << "emitter '" << m_name << "' has been dropped." << std::endl;
}


VM::BUNDLE_UAV* EMITTER::Uav(const std::string& name)
{
    auto it = find_if(begin(m_sharedUAVs), end(m_sharedUAVs), 
        [&name](const VM::BUNDLE_UAV &uav) { return uav.name == name; });
    return it == end(m_sharedUAVs) ? nullptr : &(*it);
}


const VM::BUNDLE_UAV* EMITTER::Uav(const std::string& name) const 
{ 
    auto it = find_if(begin(m_sharedUAVs), end(m_sharedUAVs), 
        [&name](const VM::BUNDLE_UAV &uav) { return uav.name == name; });
    return it == end(m_sharedUAVs) ? nullptr : &(*it);
}


VM::BUNDLE_UAV* EMITTER::UavDeadIndices() { return Uav(FxTranslator::UAV_DEAD_INDICES); }
VM::BUNDLE_UAV* EMITTER::UavParticles() { return Uav(FxTranslator::UAV_PARTICLES); }
VM::BUNDLE_UAV* EMITTER::UavStates() { return Uav(FxTranslator::UAV_STATES); }
VM::BUNDLE_UAV* EMITTER::UavInitArguments() { return Uav(FxTranslator::UAV_SPAWN_DISPATCH_ARGUMENTS); }
VM::BUNDLE_UAV* EMITTER::UavCreationRequests() { return Uav(FxTranslator::UAV_CREATION_REQUESTS); }

const VM::BUNDLE_UAV* EMITTER::UavDeadIndices() const { return Uav(FxTranslator::UAV_DEAD_INDICES); }
const VM::BUNDLE_UAV* EMITTER::UavParticles() const { return Uav(FxTranslator::UAV_PARTICLES); }
const VM::BUNDLE_UAV* EMITTER::UavStates() const { return Uav(FxTranslator::UAV_STATES); }
const VM::BUNDLE_UAV* EMITTER::UavInitArguments() const { return Uav(FxTranslator::UAV_SPAWN_DISPATCH_ARGUMENTS); }
const VM::BUNDLE_UAV* EMITTER::UavCreationRequests() const { return Uav(FxTranslator::UAV_CREATION_REQUESTS); }

uint32_t EMITTER::GetNumParticles() const
{
    return m_capacity - UavDeadIndices()->ReadCounter();
}

void EMITTER::Reset()
{
    assert(m_capacity % m_resetBundle.numthreads.x == 0);
    // reset all available particles
    m_resetBundle.Run(m_capacity / m_resetBundle.numthreads.x);
    UavDeadIndices()->OverwriteCounter(m_capacity);
}


void EMITTER::Emit(const UNIFORMS& uniforms)
{
    m_initBundle.SetConstants(uniforms);
    m_initBundle.Run(UavInitArguments()->data.As<int>()[0]);

    m_spawnBundle.SetConstants(uniforms);
    m_spawnBundle.Run(1);
}


void EMITTER::Update(const UNIFORMS& uniforms)
{
    assert(m_capacity % m_updateBundle.numthreads.x == 0);
    m_updateBundle.SetConstants(uniforms);
    m_updateBundle.Run(m_capacity / m_updateBundle.numthreads.x);
}


void EMITTER::Prerender(const UNIFORMS& uniforms) 
{
    for (int i = 0; i < m_passes.size(); ++ i)
    {
        m_passes[i].Prerender(uniforms);
    }
}


void EMITTER::Serialize() 
{
    for (int i = 0; i < m_passes.size(); ++ i)
    {
        m_passes[i].Serialize();
    }
}

void EMITTER::Simulate(const UNIFORMS& uniforms)
{
    Update(uniforms);
    Emit(uniforms);
    // dump();
}


//
// hack to allow hot reload for similar emitters
//

bool CompareTypeLayout(const Fx::TypeLayoutT& lhs, const Fx::TypeLayoutT& rhs);
bool CompareTypeField(const Fx::TypeFieldT& lhs, const Fx::TypeFieldT& rhs)
{
    if (lhs.name != rhs.name || lhs.padding != rhs.padding || lhs.size != rhs.size) 
        return false;
    return CompareTypeLayout(*lhs.type, *rhs.type);
}

bool CompareTypeLayout(const Fx::TypeLayoutT& lhs, const Fx::TypeLayoutT& rhs)
{
    if (lhs.name != rhs.name || lhs.length != rhs.length || lhs.size != rhs.size) 
        return false;
    for (int i = 0; i < lhs.fields.size(); ++i)
    {
        if (!CompareTypeField(*lhs.fields[i], *rhs.fields[i]))
            return false;
    }
    return true;
}

bool EMITTER::Copy(const EMITTER& src)
{
    if (!(operator == (src))) return false; 
    for (auto& uav : m_sharedUAVs)
    {
        memcpy(uav.buffer.As<void>(), src.Uav(uav.name)->buffer.As<void>(), uav.buffer.ByteLength());
    }
    return true;
}

bool EMITTER::operator ==(const EMITTER& rhs) const
{
    if (m_capacity != rhs.m_capacity ||
        m_passes.size() != rhs.m_passes.size() ||
        !CompareTypeLayout(m_particle, rhs.m_particle)) return false;
    for (int i = 0; i < m_passes.size(); ++i) {
        auto& lp = m_passes[i].GetDesc();
        auto& rp = rhs.m_passes[i].GetDesc();
        if (lp.geometry != rp.geometry || lp.sorting != rp.sorting) return false;
        if (!CompareTypeLayout(lp.renderInstance, rp.renderInstance)) return false;
    }
    return true;
}

//
// -- end of hack
//

// struct Part
// {
// 	vec3 speed;
// 	vec3 pos;
// 	float size;
// 	float timelife;
// };

void EMITTER::Dump()
{
    auto npart = GetNumParticles();
    auto partSize = m_particle.size;

    std::cout << "particles total: " << npart << " ( " << UavDeadIndices()->ReadCounter() / m_capacity << " )" << std::endl;

    for (int iPart = 0; iPart < UavStates()->data.size; ++ iPart)
    {
        auto alive = !!UavStates()->data[iPart];
        if (alive)
        {
            // auto& part = uavParticles()->data.As<Part>()[iPart];
            // std::cout << "part(" << iPart 
            // << ") = { size: " << part.size 
            // << ", timelife: " << part.timelife 
            // << ", pos: " << part.pos.x << ", " << part.pos.y << ", " << part.pos.z 
            // << " } " << std::endl;
        }
    }
}


}

File: src/lib/fx/emitter/ts/emitter.ts
import { assert, verbose } from '@lib/common';
import * as VM from '@lib/fx/bytecode/VM';
import { FxTranslator } from '@lib/fx/translators/FxTranslator';
import { BundleT, EPartRenderRoutines, EPartSimRoutines, PartBundleT, RoutineBytecodeBundleT, RoutineGLSLBundleT, TypeLayoutT, UAVBundleT, UIControlT } from '@lib/idl/bundles/FxBundle_generated';
import * as Bytecode from "@lib/idl/bytecode";
import {  Uniforms, IEmitter, IEmitterPass } from '@lib/idl/emitter';
import { Vector3 } from 'three';
import { asBundleMemory } from '@lib/fx/bytecode/VM/ts/bundle';
import * as FxBundle from '@lib/fx/bundles/Bundle';


type IUAVResource = ReturnType<typeof VM.createUAV>;

function createUAVEx(bundle: UAVBundleT, capacity: number): IUAVResource {
    const uav = VM.createUAV(<string>bundle.name, bundle.stride, capacity, bundle.slot);
    // console.log(`UAV '${uav.name}' (counter value: ${UAV.readCounter(uav)}, size: ${uav.length}) has been created.`);
    return uav;
}

// tslint:disable-next-line:max-line-length
// !attention! updtae shared list if needed
function createUAVsEx(bundles: UAVBundleT[], capacity: number, sharedUAVs: IUAVResource[] = []): IUAVResource[] {
    return bundles.map(uavBundle => {
        const sharedUAV = sharedUAVs.find(uav => uav.name === uavBundle.name);
        if (sharedUAV) return sharedUAV;
        const uav = createUAVEx(uavBundle, capacity);
        sharedUAVs.push(uav);
        return uav;
    });
}


function setupFxRoutineBytecodeBundle(debugName: string, routineBundle: RoutineBytecodeBundleT, capacity: number, sharedUAVs: IUAVResource[]) {
    const codeLength = routineBundle.code.length;
    if (codeLength == 0) {
        // it's dummy bundle
        return null;
    }

    const vmBundle = VM.make(debugName, routineBundle.code);
    const uavs = createUAVsEx(routineBundle.resources.uavs, capacity, sharedUAVs);
    const numthreads = routineBundle.numthreads;

    // setup VM inputs
    uavs.forEach(uav => { vmBundle.setInput(uav.index, uav.buffer); });

    function setConstants(constants: Uniforms) {
        Object.keys(constants)
            .forEach(name => vmBundle.setConstant(name, constants[name]));
    }

    assert(numthreads[0] >= 1 && numthreads[1] === 1 && numthreads[2] === 1);

    function run(numgroups: number) {
        vmBundle.dispatch({ x: numgroups, y: 1, z: 1 }, { x: numthreads[0], y: numthreads[1], z: numthreads[2] });
    }

    return {
        uavs,
        bundle: vmBundle,
        run,
        setConstants,
        groupsizex: numthreads[0]
    };
}


const UAV = {
    overwriteCounter(uav: Bytecode.IUAV, value: number) {
        VM.memoryToI32Array(uav.buffer)[0] = value;
    },

    readCounter(uav: Bytecode.IUAV): number {
        return VM.memoryToI32Array(uav.buffer)[0];
    },

    readElement({ data, elementSize }: Bytecode.IUAV, iElement: number): Uint8Array {
        const u8a = VM.memoryToU8Array(data);
        return new Uint8Array(u8a.buffer, u8a.byteOffset + iElement * elementSize, elementSize);
    },

    minidump(uav: Bytecode.IUAV): void
    {
        const { name, length, elementSize, register, data } = uav;
        // std::cout << "--------------------------------------" << std::endl;
        console.log(` uav ${name}[${length}x${elementSize}:r${register}:cnt(${UAV.readCounter(uav)})]`);
        
        const u8a = VM.memoryToU8Array(data);
        let n = Math.min(64, length * elementSize);
        let sout = '';
        for (let i = 0; i < n; ++ i)
        {
            sout += `${u8a[i].toString(16)} `;
        }
        sout += '...';
        console.log(sout);
        // std::cout << "--------------------------------------" << std::endl;
    }
};



// tslint:disable-next-line:max-func-body-length
function createEmiterFromBundle(bundle: BundleT, uavResources: IUAVResource[]): IEmitter {
    const { name, content } = bundle;
    const { capacity, particle, simulationRoutines, renderPasses } = content as PartBundleT;

    const resetBundle = setupFxRoutineBytecodeBundle(`${name}/reset`, <RoutineBytecodeBundleT>simulationRoutines[EPartSimRoutines.k_Reset], capacity, uavResources);
    const initBundle = setupFxRoutineBytecodeBundle(`${name}/init`, <RoutineBytecodeBundleT>simulationRoutines[EPartSimRoutines.k_Init], capacity, uavResources);
    const updateBundle = setupFxRoutineBytecodeBundle(`${name}/update`, <RoutineBytecodeBundleT>simulationRoutines[EPartSimRoutines.k_Update], capacity, uavResources);
    const spawnBundle = setupFxRoutineBytecodeBundle(`${name}/spawn`, <RoutineBytecodeBundleT>simulationRoutines[EPartSimRoutines.k_Spawn], 4, uavResources);

    const uavDeadIndices = uavResources.find(uav => uav.name === FxTranslator.UAV_DEAD_INDICES);
    const uavParticles = uavResources.find(uav => uav.name === FxTranslator.UAV_PARTICLES);
    const uavStates = uavResources.find(uav => uav.name === FxTranslator.UAV_STATES);
    const uavInitArguments = uavResources.find(uav => uav.name === FxTranslator.UAV_SPAWN_DISPATCH_ARGUMENTS);
    const uavCreationRequests = uavResources.find(uav => uav.name === FxTranslator.UAV_CREATION_REQUESTS);

    function preparePrerender() {
        passes.forEach((p, i) => {
            const uavPrerendered = uavResources.find(uav => uav.name === `${FxTranslator.UAV_PRERENDERED}${i}`);
            const uavSerials = uavResources.find(uav => uav.name === `${FxTranslator.UAV_SERIALS}${i}`);
            if (uavPrerendered) {
                UAV.overwriteCounter(uavPrerendered, 0);
            }
            if (uavSerials) {
                UAV.overwriteCounter(uavSerials, 0);
            }
        });
    }

    const passes = renderPasses.map((pass, i) => {
        const {
            routines,
            geometry,
            sorting,
            instanceCount,
            instance,
            stride
        } = pass;

        const UAV_PRERENDERED = `${FxTranslator.UAV_PRERENDERED}${i}`;
        const UAV_SERIALS = `${FxTranslator.UAV_SERIALS}${i}`;

        const prerenderBundle = routines[EPartRenderRoutines.k_Prerender];
        const bundle = setupFxRoutineBytecodeBundle(`${name}/prerender`, <RoutineBytecodeBundleT>prerenderBundle, capacity * instanceCount, uavResources);
        
        const uavPrerendered = uavResources.find(uav => uav.name === UAV_PRERENDERED);
        const uavSerials = uavResources.find(uav => uav.name === UAV_SERIALS);

        const vertexShader = <string>routines[EPartRenderRoutines.k_Vertex].code;
        const pixelShader = <string>routines[EPartRenderRoutines.k_Pixel].code;

        // note: only GLSL routines are supported!
        const instanceLayout = (<RoutineGLSLBundleT>routines[EPartRenderRoutines.k_Vertex]).attributes;
        const getNumRenderedParticles = () => UAV.readCounter(uavPrerendered) * instanceCount;

        // if no prerender bundle then all particles must be prerendered within update stage
        // looking for prerendered reflection among prerender or update routine uavs
        const uavPrerendReflect: UAVBundleT = (<RoutineBytecodeBundleT>(bundle ? routines[EPartRenderRoutines.k_Prerender] : simulationRoutines[EPartSimRoutines.k_Update]))
            .resources.uavs.find(uavReflection => uavReflection.name === UAV_PRERENDERED);

        //
        // Sorting
        //

        const uavNonSorted = uavPrerendered;
        const uavNonSortedU8 = VM.memoryToU8Array(uavNonSorted.data);

        let uavPrerendReflectSorted: UAVBundleT = null;
        let uavSorted: Bytecode.IUAV = null;
        let uavSortedU8: Uint8Array = null;
        let uavSerialsI32: Int32Array = null;
        
        if (sorting)
        {
            uavPrerendReflectSorted = new UAVBundleT(`${uavPrerendReflect.name}Sorted`, uavPrerendReflect.slot, uavPrerendReflect.stride, uavPrerendReflect.type);
            uavSorted = createUAVsEx([uavPrerendReflectSorted], capacity * instanceCount, uavResources)[0];
            uavSortedU8 = VM.memoryToU8Array(uavSorted.data);
            uavSerialsI32 = VM.memoryToI32Array(uavSerials.data);
        }
        

        // dump prerendered particles
        const dump = (): void => {
            let nPart = getNumRenderedParticles();
            verbose(`dump ${nPart}/${capacity} prerendred particles: `);
            for (let iElement = 0; iElement < nPart; ++iElement) {
                verbose(VM.asNativeRaw(UAV.readElement(uavNonSorted, iElement), instance));
            }
        };

        function serialize() {
            if (!sorting) {
                return;
            }

            const nStrideF32 = stride * instanceCount; // stride in floats

            assert(uavSortedU8.byteLength >> 2 === nStrideF32 * capacity);

            const srcF32 = new Float32Array(uavNonSortedU8.buffer, uavNonSortedU8.byteOffset, uavNonSortedU8.byteLength >> 2);
            const dstF32 = new Float32Array(uavSortedU8.buffer, uavSortedU8.byteOffset, uavSortedU8.byteLength >> 2);

            const indicies = [];

            // todo: sort inplace using serials pairs
            for (let iPart = 0; iPart < UAV.readCounter(uavPrerendered); ++iPart) 
            {
                const sortIndex = uavSerialsI32[iPart * 2 + 0];
                const partIndex = uavSerialsI32[iPart * 2 + 1];
                indicies.push([ partIndex, sortIndex ]);
            };
            indicies.sort((a, b) => -a[1] + b[1]);

            for (let i = 0; i < indicies.length; ++i) {
                const iFrom = indicies[i][0] * nStrideF32;
                const iTo = i * nStrideF32;

                const from = srcF32.subarray(iFrom, iFrom + nStrideF32);
                const copyTo = dstF32.subarray(iTo, iTo + nStrideF32);
                copyTo.set(from);
            }
        }

        function getData() { return asBundleMemory(sorting ? uavSortedU8 : uavNonSortedU8); }
        function getDesc() {
            return {
                instanceName: instance.name as string,
                instanceLayout: instanceLayout.map(({ name, offset, size }) => ({ name: <string>name, offset, size })), // FIXME
                stride,
                geometry: <string>geometry,                                                                             // FIXME
                sorting,
                vertexShader,
                pixelShader,
            };
        }


        function prerender(uniforms: Uniforms)
        {
            if (!bundle) {
                // manual prerender is used
                return;
            }

            bundle.setConstants(uniforms);
            bundle.run(Math.ceil(capacity / bundle.groupsizex));
        }

        return {
            getDesc,
            getNumRenderedParticles,                                                                           // FIXME
            getData,
            serialize,
            prerender,
            dump
        };
    });

    const getNumParticles = () => capacity - UAV.readCounter(uavDeadIndices);
    const getName = () => <string>name;
    const getType = (): 'emitter' => 'emitter';
    const getPassCount = () => passes.length;
    const getPass = (i: number) => passes[i];
    const getCapacity = () => capacity;

    function reset() {
        // reset all available particles
        resetBundle.run(Math.ceil(capacity / resetBundle.groupsizex));
        UAV.overwriteCounter(uavDeadIndices, capacity);
    }


    function update(uniforms: Uniforms) {
        preparePrerender();
        updateBundle.setConstants(uniforms);
        updateBundle.run(Math.ceil(capacity / updateBundle.groupsizex));
    }


    function prerender(uniforms: Uniforms) {
        preparePrerender();
        passes.forEach(pass => pass.prerender(uniforms));
    }

    function serialize() {
        passes.forEach(pass => pass.serialize());
    }


    function emit(uniforms: Uniforms) {
        initBundle.setConstants(uniforms);
        initBundle.run(VM.memoryToI32Array(uavInitArguments.data)[0]);

        spawnBundle.setConstants(uniforms);
        spawnBundle.run(1);
    }

    function simulate(uniforms: Uniforms)
    {
        update(uniforms);
        emit(uniforms);
    }

    function dump() {
        const npart = getNumParticles();
        const partSize = particle.size;

        verbose(`particles total: ${npart} ( ${UAV.readCounter(uavDeadIndices)}/${capacity} )`);

        const uavStatesI32 = VM.memoryToI32Array(uavStates.data);
        const uavParticlesU8 = VM.memoryToU8Array(uavParticles.data);

        uavStatesI32.forEach((alive, iPart) => {
            if (alive) {
                const partRaw = new Uint8Array(uavParticlesU8.buffer, uavParticlesU8.byteOffset + iPart * partSize, partSize);
                verbose(iPart, VM.asNativeRaw(partRaw, particle));
            }
        });
    }

    return {
        // abstract interface
        getType,
        getName,
        getCapacity,
        getPassCount,
        getPass,
        getNumParticles,
        
        reset,
        simulate,
        prerender,
        serialize,
        
        dump
    };
}

type TSEmitter = ReturnType<typeof createTsEmitter>;

function compareFxTypeLayouts(left: TypeLayoutT, right: TypeLayoutT) {
    return JSON.stringify(left) == JSON.stringify(right);
}

// function compareFxControls(left: UIControlT[], right: UIControlT[]) {
//     return JSON.stringify(left) == JSON.stringify(right);
// }

// todo: rework comparisson to be more readable and compact
function comparePartFxBundles(left: PartBundleT, right: PartBundleT): boolean {
    if (left.capacity != right.capacity) return false;
    if (left.renderPasses.length != right.renderPasses.length) return false;
    if (!compareFxTypeLayouts(left.particle, right.particle)) return false;
    for (let i = 0; i < left.renderPasses.length; ++i) {
        if (left.renderPasses[i].geometry != right.renderPasses[i].geometry) return false;
        if (left.renderPasses[i].sorting != right.renderPasses[i].sorting) return false;
        if (left.renderPasses[i].instanceCount != right.renderPasses[i].instanceCount) return false;
        if (!compareFxTypeLayouts(left.renderPasses[i].instance, right.renderPasses[i].instance)) return false;
    }
    return true;
}

export function copyTsEmitter(dst: IEmitter, src: IEmitter): boolean
{
    if (comparePartFxBundles((<TSEmitter>dst).bundle.content as PartBundleT, (<TSEmitter>src).bundle.content as PartBundleT)) 
    {
        (<TSEmitter>dst).uavResources.forEach((uav, i) =>
            VM.memoryToU8Array(uav.buffer).set(VM.memoryToU8Array((<TSEmitter>src).uavResources[i].buffer)));
        return true;
    }
    return false;
}

export function destroyTsEmitter(emitter: IEmitter): void 
{
    let { uavResources } = <TSEmitter>emitter;
    uavResources.forEach(uav => {
        VM.destroyUAV(uav);
        // verbose(`UAV '${uav.name}' has been destroyed.`);
    });
    verbose(`emitter '${name}' has been dropped.`);
}

export function createTsEmitter(bundle: BundleT)
{
    let uavResources: Bytecode.IUAV[] = [];
    let newly = createEmiterFromBundle(bundle, uavResources);
    return { bundle, uavResources, ...newly };
}
